File: src/leetCodeProblems/MathCalculations/AddStrings415.java
package leetCodeProblems.MathCalculations;

/**
 * LeetCode - https://leetcode.com/problems/add-strings/
 * Time Complexity - O(n)
 * Space Complexity - O(1)
 */
public class AddStrings415 {

    public String addStrings(String num1, String num2) {

        int num1Pointer = num1.length()-1;
        int num2Pointer = num2.length()-1;
        int carry = 0;

        int localOutputNumb = 0;
        int localNum1Val = 0;
        int localNum2Val = 0;

        String output = "";

        while (num1Pointer >= 0 || num2Pointer >= 0) {

            localNum1Val = 0;
            localNum2Val = 0;

            if (num1Pointer >= 0) {
                localNum1Val = Integer.parseInt(String.valueOf(num1.charAt(num1Pointer)));
            }

            if (num2Pointer >= 0) {
                localNum2Val = Integer.parseInt(String.valueOf(num2.charAt(num2Pointer)));
            }

            localOutputNumb = localNum1Val + localNum2Val + carry;

            output = localOutputNumb%10 + output;
            carry = localOutputNumb/10;

            num1Pointer--;
            num2Pointer--;
        }

        if (carry != 0) {
            output = carry + output;
        }

        return output;
    }

    public static void main(String[] args) {

        String num1 = "11";
        String num2 = "123";
        AddStrings415 obj = new AddStrings415();
        //System.out.println(obj.addStrings(num1, num2)); // expected o/p = 134

        num1 = "456";
        num2 = "77";

        System.out.println(obj.addStrings(num1, num2)); // expected o/p = 533

        //num1 = "0";
        //num2 = "0";

        //System.out.println(obj.addStrings(num1, num2)); // expected o/p = 0

    }
}

File: src/leetCodeProblems/MathCalculations/AddStrings415.go
package main

import (
	"log"
	"strconv"
)

/*
- LeetCode - https://leetcode.com/problems/add-strings/
- Time - O(n)
- Space - O(1)
*/

func addStrings(num1 string, num2 string) string {

	carry := 0
	out := ""
	i1 := len(num1) - 1
	i2 := len(num2) - 1

	for i1 >= 0 || i2 >= 0 {

		sum := carry

		if i1 >= 0 {
			num, _ := strconv.Atoi(string(num1[i1]))
			sum += num
		}

		if i2 >= 0 {
			num, _ := strconv.Atoi(string(num2[i2]))
			sum += num
		}

		out = strconv.Itoa(sum%10) + out
		carry = sum / 10
		i2--
		i1--
	}

	if carry > 0 {
		out = strconv.Itoa(carry) + out
	}
	return out
}

func main() {

	log.Println(addStrings("11", "123"))
	log.Println(addStrings("456", "77"))
}
